packages/ember-metal/lib/meta.js

```diff
@@ -104,11 +117,70 @@ Meta.prototype.isInitialized = function(obj) {
   return this.proto !== obj;
 };
 
+const NODE_STACK = [];
+
+Meta.prototype.destroy = function() {
+  // remove chainWatchers to remove circular references that would prevent GC
+
+  let node, nodes, key, nodeObject;
+  node = this.readableChains();
+  if (node) {
+    NODE_STACK.push(node);
+    // process tree
+    while (NODE_STACK.length > 0) {
+      node = NODE_STACK.pop();
+      // push children
+      nodes = node._chains;
+      if (nodes) {
+        for (key in nodes) {
+          if (nodes[key] !== undefined) {
+            NODE_STACK.push(nodes[key]);
+          }
+        }
+      }
+
+      // remove chainWatcher in node object
+      if (node._watching) {
+        nodeObject = node._object;
+        if (nodeObject) {
+          let foreignMeta = peekMeta(nodeObject);
+          // avoid cleaning up chain watchers when both current and
+          // foreign objects are being destroyed
```

Comment:
just comment here that marking a source as destroying means you are expected to not be holding onto a reference to the source from anything reachable.